File: ch23p810textFieldDelegate/ch23p810textFieldDelegate/MyTextField.m


#import "MyTextField.h"

// ??? is this legal ???

/*

@interface UITextField (SelectionLikeTextView)
- (NSRange) selectedRange;
- (void) setSelectedRange: (NSRange) range;
@end

@implementation UITextField (SelectionLikeTextView)
- (NSRange) selectedRange {
    UITextRange* r = self.selectedTextRange;
    NSInteger loc = [self offsetFromPosition:self.beginningOfDocument
                                  toPosition:r.start];
    NSInteger len = [self offsetFromPosition:r.start
                                  toPosition:r.end];
    return NSMakeRange(loc,len);
}
- (void) setSelectedRange: (NSRange) r {
    UITextPosition* st = [self positionFromPosition:self.beginningOfDocument
                                             offset:r.location];
    UITextPosition* en = [self positionFromPosition:self.beginningOfDocument
                                             offset:r.location + r.length];
    [self setSelectedTextRange:[self textRangeFromPosition:st toPosition:en]];
}
@end
 
 */


@implementation MyTextField

+ (NSString*) stateForAbbrev: (NSString*) abbrev {
    static NSArray* list = nil;
    if (nil == list) {
        NSURL* path = [[NSBundle mainBundle] URLForResource:@"abbreviations" withExtension:@"txt"];
        NSString* s = [NSString stringWithContentsOfURL:path encoding:NSUTF8StringEncoding error:nil];
        list = [s componentsSeparatedByString:@"\n"];
    }
    NSUInteger ix = [list indexOfObject:[abbrev uppercaseString]];
    if (NSNotFound == ix)
        return nil;
    return list[ix+1];
}

// hard to see why this would be strictly illegal

- (BOOL) canPerformAction:(SEL)action withSender: (id) sender {
    if (action == @selector(expand:)) {
        NSString* s = [self textInRange:self.selectedTextRange];
        return (s.length == 2 && [self.class stateForAbbrev: s]);
    }
    return [super canPerformAction:action withSender:sender];
}

- (void) expand: (id) sender {
    NSString* s = [self textInRange:self.selectedTextRange];
    s = [self.class stateForAbbrev:s];
    [self replaceRange:self.selectedTextRange withText:s];
}

- (void) copy: (id) sender {
    [super copy: sender];
    UIPasteboard* pb = [UIPasteboard generalPasteboard];
    NSString* s = pb.string;
    // ... alter s here ....
    s = [s stringByAppendingString:@"suprise!"];
    pb.string = s;
}

@end

File: iOS8bookExamples/bk2ch10p526textFieldDelegate/ch23p810textFieldDelegate/MyTextField.swift


import UIKit


class MyTextField: UITextField {
    
    let list : [String] = {
        let path = NSBundle.mainBundle().URLForResource("abbreviations", withExtension:"txt")!
        let s = String(contentsOfURL:path, encoding:NSUTF8StringEncoding, error:nil)!
        return s.componentsSeparatedByString("\n")
        }()
    
    func stateForAbbrev(abbrev:String) -> String? {
        let ix = find(self.list, abbrev.uppercaseString)
        return ix != nil ? list[ix!+1] : nil
    }
    
    
    override func canPerformAction(action: Selector, withSender sender: AnyObject?) -> Bool {
        if action == "expand:" {
            if let r = self.selectedTextRange {
                let s = self.textInRange(r)
                return count(s) == 2 && self.stateForAbbrev(s) != nil
            }
            
        }
        return super.canPerformAction(action, withSender:sender)
    }
    
    func expand(sender:AnyObject?) {
        if let r = self.selectedTextRange {
            let s = self.textInRange(r)
            if let ss = self.stateForAbbrev(s) {
                self.replaceRange(r, withText:ss)
            }
        }
    }
    
    override func copy(sender:AnyObject?) {
        super.copy(sender)
        let pb = UIPasteboard.generalPasteboard()
        if let s = pb.string {
            // ... alter s here ...
            let ss = s + "surprise!"
            pb.string = ss
        }
    }
}

extension UITextField { // is this legal? sure makes life better...
    var selectedRange : NSRange {
    get {
        if let r = self.selectedTextRange {
            let loc = self.offsetFromPosition(self.beginningOfDocument,
                toPosition:r.start)
            let len = self.offsetFromPosition(r.start,
                toPosition:r.end)
            return NSMakeRange(loc,len)
        } else {
            return NSMakeRange(0,0)
        }
    }
    set (r) {
        let st = self.positionFromPosition(self.beginningOfDocument,
            offset:r.location)
        let en = self.positionFromPosition(self.beginningOfDocument,
            offset:r.location + r.length)
        self.selectedTextRange = self.textRangeFromPosition(st, toPosition:en)
    }
    }
}


